File: frontend/src/shared/utils/videoEmbed.js
import { getSafeExternalUrl } from './linkSafety.js';

export function normalizeVideoUrl(url) {
  const raw = String(url || '').trim();
  if (!raw) return '';
  if (/^[A-Za-z0-9_-]{11}$/.test(raw)) return `https://youtu.be/${raw}`;
  if (/^\/\//.test(raw)) return `https:${raw}`;
  if (/^[a-z][a-z\d+.-]*:/i.test(raw)) {
    try {
      const parsed = new URL(raw);
      return ['http:', 'https:'].includes(parsed.protocol) ? parsed.toString() : '';
    } catch {
      return '';
    }
  }
  if (/^(www\.|youtu\.be\/|youtube\.com\/|m\.youtube\.com\/|vimeo\.com\/|player\.vimeo\.com\/|drive\.google\.com\/)/i.test(raw)) {
    return `https://${raw}`;
  }
  return raw;
}

function normalizeStartSeconds(value) {
  const seconds = Math.floor(Number(value || 0));
  if (!Number.isFinite(seconds) || seconds <= 0) return 0;
  return Math.min(seconds, 23 * 60 * 60 + 59 * 60 + 59);
}

function withParams(src, params) {
  const u = new URL(src);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null && value !== '') {
      u.searchParams.set(key, value);
    }
  }
  return u.toString();
}

function currentOrigin() {
  if (typeof window === 'undefined' || !window.location?.origin) return '';
  return window.location.origin;
}

export function getVideoEmbed(url, options = {}) {
  const raw = normalizeVideoUrl(url);
  if (!raw) return null;
  const safeExternalUrl = getSafeExternalUrl(raw);
  const startSeconds = normalizeStartSeconds(options.startSeconds);
  const youtubeParams = {
    rel:'0',
    modestbranding:'1',
    playsinline:'1',
    iv_load_policy:'3',
    cc_load_policy:'1',
    disablekb:'0',
    controls:'1',
    fs:'1',
    origin:currentOrigin(),
    start:startSeconds || undefined,
  };
  try {
    const u = new URL(raw);
    if (!['http:', 'https:'].includes(u.protocol)) return null;
    const host = u.hostname.replace(/^www\./, '');
    if (host === 'youtu.be') {
      const id = u.pathname.split('/').filter(Boolean)[0];
      if (id) return { type:'iframe', src:withParams(`https://www.youtube.com/embed/${id}`, youtubeParams), provider:'youtube' };
    }
    if (host.includes('youtube.com')) {
      const watchId = u.searchParams.get('v');
      const path = u.pathname.split('/').filter(Boolean);
      const embeddedId = ['embed', 'shorts', 'live'].includes(path[0]) ? path[1] : null;
      const id = watchId || embeddedId;
      if (id) return { type:'iframe', src:withParams(`https://www.youtube.com/embed/${id}`, youtubeParams), provider:'youtube' };
    }
    if (host.includes('vimeo.com')) {
      const id = u.pathname.split('/').filter(Boolean).find(part => /^\d+$/.test(part));
      if (id) {
        const src = withParams(`https://player.vimeo.com/video/${id}`, { dnt:'1', title:'0', byline:'0', portrait:'0', badge:'0' });
        return { type:'iframe', src: startSeconds ? `${src}#t=${startSeconds}s` : src, provider:'vimeo' };
      }
    }
    if (host === 'drive.google.com') {
      const path = u.pathname.split('/').filter(Boolean);
      const fileIndex = path.indexOf('d');
      const id = fileIndex >= 0 ? path[fileIndex + 1] : u.searchParams.get('id');
      if (id) return { type:'iframe', src:`https://drive.google.com/file/d/${id}/preview` };
    }
  } catch {
    // Fall through to direct media detection.
  }
  if (safeExternalUrl && /\.(mp4|webm|ogg|mov)(?:[?#].*)?$/i.test(safeExternalUrl)) return { type:'video', src:safeExternalUrl };
  return { type:'blocked', externalUrl:safeExternalUrl };
}

File: mobile/lib/features/lessons/video_embed.dart
// Dart port of frontend/src/shared/utils/videoEmbed.js
// Handles YouTube, Vimeo, Google Drive, and direct MP4/WebM/OGG/MOV URLs.

enum VideoEmbedType { iframe, video, blocked, none }

class VideoEmbed {
  final VideoEmbedType type;
  final String src;
  final String provider; // 'youtube' | 'vimeo' | 'drive' | ''
  // True for a vertical video (e.g. a YouTube Short) — the player should use
  // a 9:16 box instead of the default 16:9, or the picture gets stretched.
  final bool isVertical;
  const VideoEmbed({
    required this.type,
    this.src = '',
    this.provider = '',
    this.isVertical = false,
  });
}

String _normalizeUrl(String raw) {
  final s = raw.trim();
  if (s.isEmpty) return '';
  // Raw 11-char YouTube ID
  if (RegExp(r'^[A-Za-z0-9_-]{11}$').hasMatch(s)) return 'https://youtu.be/$s';
  // Protocol-relative
  if (s.startsWith('//')) return 'https:$s';
  // Has a scheme — validate http/https only
  if (RegExp(r'^[a-z][a-z\d+.\-]*:', caseSensitive: false).hasMatch(s)) {
    try {
      final u = Uri.parse(s);
      return (u.scheme == 'http' || u.scheme == 'https') ? s : '';
    } catch (_) {
      return '';
    }
  }
  // Bare hostnames / partial URLs
  if (RegExp(
    r'^(www\.|youtu\.be\/|youtube\.com\/|m\.youtube\.com\/|vimeo\.com\/|player\.vimeo\.com\/|drive\.google\.com\/)',
    caseSensitive: false,
  ).hasMatch(s)) {
    return 'https://$s';
  }
  return s;
}

VideoEmbed getVideoEmbed(String url) {
  final raw = _normalizeUrl(url);
  if (raw.isEmpty) return const VideoEmbed(type: VideoEmbedType.none);

  Uri u;
  try {
    u = Uri.parse(raw);
    if (u.scheme != 'http' && u.scheme != 'https') {
      return const VideoEmbed(type: VideoEmbedType.none);
    }
  } catch (_) {
    return const VideoEmbed(type: VideoEmbedType.none);
  }

  final host = u.host.replaceFirst(RegExp(r'^www\.'), '');
  final segments = u.pathSegments.where((s) => s.isNotEmpty).toList();

  // ── YouTube ────────────────────────────────────────────────────────────────
  if (host == 'youtu.be') {
    final id = segments.isNotEmpty ? segments.first : '';
    if (id.isNotEmpty) return _ytEmbed(id);
  }
  if (host.contains('youtube.com')) {
    final first = segments.isNotEmpty ? segments.first : '';
    final watchId = u.queryParameters['v'];
    final embeddedId = ['embed', 'shorts', 'live'].contains(first)
        ? (segments.length > 1 ? segments[1] : null)
        : null;
    final id = watchId ?? embeddedId ?? '';
    if (id.isNotEmpty) return _ytEmbed(id, isVertical: first == 'shorts');
  }

  // ── Vimeo ──────────────────────────────────────────────────────────────────
  if (host.contains('vimeo.com')) {
    final id = segments.firstWhere(
      (s) => RegExp(r'^\d+$').hasMatch(s),
      orElse: () => '',
    );
    if (id.isNotEmpty) {
      return VideoEmbed(
        type: VideoEmbedType.iframe,
        src: 'https://player.vimeo.com/video/$id?dnt=1&title=0&byline=0&portrait=0&badge=0',
        provider: 'vimeo',
      );
    }
  }

  // ── Google Drive ───────────────────────────────────────────────────────────
  // Back to Drive's `/preview` page (their own file viewer + player chrome).
  // The "clean" direct-stream URL (`uc?export=download`) only works when a
  // file is fully public and small enough to skip Google's confirmation
  // flow — for a real private/shared file it needs cookies and an
  // interactive confirmation that a bare <video> tag can't do, so it just
  // fails to load. `/preview` is Google's own page handling all of that
  // internally, so it reliably plays — the trade-off is Drive's own chrome
  // (open/download icons) stays visible, same as Vimeo's iframe below.
  if (host == 'drive.google.com') {
    final dIdx = segments.indexOf('d');
    final id = (dIdx >= 0 && dIdx + 1 < segments.length)
        ? segments[dIdx + 1]
        : u.queryParameters['id'] ?? '';
    if (id.isNotEmpty) {
      return VideoEmbed(
        type: VideoEmbedType.iframe,
        src: 'https://drive.google.com/file/d/$id/preview',
        provider: 'drive',
      );
    }
  }

  // ── Direct video file ──────────────────────────────────────────────────────
  if (RegExp(r'\.(mp4|webm|ogg|mov)([?#].*)?$', caseSensitive: false).hasMatch(raw)) {
    return VideoEmbed(type: VideoEmbedType.video, src: raw);
  }

  return VideoEmbed(type: VideoEmbedType.blocked, src: raw);
}

VideoEmbed _ytEmbed(String id, {bool isVertical = false}) => VideoEmbed(
      type: VideoEmbedType.iframe,
      src: 'https://www.youtube.com/embed/$id'
          '?rel=0&modestbranding=1&playsinline=1&iv_load_policy=3&controls=1&fs=1',
      provider: 'youtube',
      isVertical: isVertical,
    );
